### ISSUE TITLE
support MongoDB

### ISSUE THREAD

**Issue by freeznet** [kuzzleio/kuzzle#364]
Hi there, do you have any plan to add MongoDB support? if not, i would like to do it by myself, but seems the `lib/service/README` document is out-dated, so is there any updated guide?

**Comment by benoitvidis** [comment]
Hi. Indeed, the README is not up-to-date anymore, sorry about that... Sorry for noticing, we are going to update it soon.

On the current version (RC5), a service must implement two methods: 

1. A constructor, taking 2 arguments for input: `kuzzle` and `options`. For the time being, the `options` object is almost empty but contains a property `service` which is the name of the service (`ReadEngine`, `WriteEngine`, etc).
2. An `init` method, wich does not take any arguments but must return a Promise. This method is called during Kuzzle start.

In your case, I suppose you want to override one of Kuzzle data service. If so, you'll need to implement all methods that are defined in [elasticsearch.js service file](https://github.com/kuzzleio/kuzzle/blob/master/lib/services/elasticsearch.js).

The configuration to tell Kuzzle to use one service or another is somewhere in lib/config, for the moment using some unneeded over-complicated process (we are also working on simplifying this part).

The tricky part is that some internal data for Kuzzle are persisted using something else than a service, namely the [`internalEngine`](https://github.com/kuzzleio/kuzzle/blob/master/lib/services/internalEngine.js), which uses a more direct link to Elasticsearch and uses some slightly different method signatures.

If your target is to totally get rid of Elasticsearch, you also will need to create something similar using MongoDb. For this last one, there is no configuration to tell Kuzzle to use something else than the default engine and you'll probably need to implement it on your side.

Finally, we are currently in the process of debating how this can simplify the interaction between Kuzzle and its data layer.
One thing is to unify the 3 engines we currently have (`internalEngine`, `readEngine`, `writeEngine`).

Another point is about extensibility and flexibility. One option which is being discussed would be to accept the fact Kuzzle has some strong dependency on Elasticsearch and to drop some ability to replace the data services, which would make what you want to achieve more difficult.

May I ask what your use case is? If there is some need to keep some flexibility in Kuzzle data storage, we definitely need to think twice about being too coupled with ES and you input about would be very valuable to us.

**Comment by farwayer** [comment]
Hi and thank you! kuzzle is really nice. But database support is a weakness of framework.
> Elasticsearch is a search engine and not a database.

This is phrase from your [blog post](http://kuzzle.io/blog/2015/11/what-nosql-solution-should-you-choose-mongodb-elastisearch-oriendb/). ES is not ACID compliant. It can lose data on hot usage. So it was very strange decision to make ES the main storage of backend.

[Discussion on official ES forum](https://discuss.elastic.co/t/elasticsearch-2-3-as-primary-data-store/50265/3)
[SO question (one year old but look like still actual)](http://stackoverflow.com/a/31378987/2175114)
[One of ES developer answer on SO (july 2016)](http://stackoverflow.com/a/38587379/2175114)

It will be very nice if you implement database abstraction layer and add database plugins support.

**Comment by scottinet** [comment]
Hi!

This is a very important and interesting question. One that requires careful thoughts before answering.

We obviously did not choose Elasticsearch on a whim, and we are well aware of its limitations. Bear in mind though that even if their developers do not recommend using ES as a primary storage solution, many of their clients do so without any problem.
That said, we also discuss (very) regularly about this potential issue, as it can have impacts for businesses, and there are multiple approaches that can be taken to address this.

Expect a longer and a lot more detailed answer next week: this question deserve attention, and we want to make sure we give you enough insight about why we made this choice for our data storage, and what are our plans for the future.

**Comment by xbill82** [comment]
First of all, thank you for your question. It really pushed us to formalizing a whole lot of thoughts that have been fluctuating among the Kuzzle team during a lot of time.
To answer your question straight to the point, **yes, ElasticSearch is not meant to be used as a so-called "master" database**. But that's not the position that Kuzzle wants to be in your architecture.
Kuzzle is meant to be the upmost, "façade", layer of your back-end. The thing that your front-end runs against, like in the schema below.

    +------+                   +--------+       +--------+
    |      |                   |        |       |        |
    |  UI  | <--- Network ---> | Kuzzle | <---> | MASTER |
    |      |                   |        |       |   DB   |
    +------+                   +--------+       +--------+

The way Kuzzle is meant to be used is as an **abstraction** of the underlying backend, acting as an **interface** between your clients and the rest of your architecture.
This is a rather common pattern in medium and big-sized architectures and that's where the power of ElasticSearch comes into play.
In order to make your front-end reactive and snappy, Kuzzle needs to be able to provide data at light speed, which is often an issue with traditional relational databases.
ElasticSearch provides the performance that we need to hydrate your client with data in the fastest way possible. That's why we think that ElasticSearch is meant to **cache your data and keep it in a shape that's easily servable to your front-end**.

Say you have a blog and want to serve a set of articles together with their comments. In your primary DB, you are used to have an `articles` and a `comments` table, which you can join associating the primary key of the comments with the comments IDs contained in every article row.
This operation is costly, and many recur to the solution of caching the joint data in a separate database.
NoSQL databases are optimized for this task, since it becomes extremely easy to store every article as a document and put its comments inside every article, skipping the need for joins.

In order to make this possible, you must ensure the consistency between your primary DB and ElasticSearch, which can be done in the following ways.

- By telling Kuzzle to commit all the data mutations both in ES and the primary DB. This can be done via the Worker Plugins, which are executed on a separate process but are triggered by Kuzzle on data-driven events.
- By periodically running a script that crushes the data in your primary DB and stores it in ElasticSearch.

Both solutions depend heavily on your data-model and on your front-end architecture. We don't think you should adapt your architecture to Kuzzle, rather the opposite. Kuzzle is meant to be plugged on your existing backend without changing it.
That's why **we plan to release a comprehensive toolkit** of examples and cookbooks **to help you easily create the Kuzzle Plugins that ensure the consistency between the ElasticSearch and your primary DB**.

In this context, you are not meant to rely entirely on ElasticSearch as a primary database, but rather store your data in a primary DB (which may also handle complex data structures like graphs) and serve it via Kuzzle (which leverages the power of ElasticSearch in the last place). We call the data in ES the "hot data", since it must be served quickly, and the data in the primary DB the "cold" one, since real-time access are less of a concern (data in this form may be more meaningful for big-data analysis, for example).

Now, that being said, we'd like to point out a few more aspects of this reflection.

- ElasticSearch is damn fast, it scales very well and it's backed by Lucene, which enables complex and fine-tuned searches. These are the main reasons why we chose ES.
- ElasticSearch is actually not ACID. But we do not seek ACID-ity here since we do not ship Kuzzle as the primary DB solution for your back-end. Since you'll be able to replicate your data between ES and your primary DB, you'll be free to choose one that ensures ACID transactions. This will not, of course, ensure the ACID-ity of all the transactions along the architecture (like if something goes wrong before your request reaches your primary DB), but the only solution to this problem would be to expose your primary DB to your clients (which is something you should definitely not do). We also do not think that ACID-ity is a goal _per se_, but rather a requirement of _some parts_ of your architecture. Using an ACID database as the primary one, you can be sure that, at every moment, your data will be consistent, even in the (very rare) case it's corrupted in ElasticSearch.
- One issue we are currently experiencing with ElasticSearch is that, between the moment you commit your mutation (basically, send a `write`) and the moment when data is actually flushed to the database, the data waits in volatile memory for around 1s (flushes happen at a 1s regular pace). For performance reasons, ES sends the `acknowledged` response right after the commit and cannot ensure the flush has been done. If ES crashes before flushing the pending data, the transaction is silently lost. This issue is currently being addressed by the Elastic Team, and the upcoming version of ElasticSearch will provide a [synchronous mode](https://www.elastic.co/guide/en/elasticsearch/reference/5.x/docs-refresh.html) (disabled by default), which will force ES to immediately flush the data to the persistent memory layer and acknowledge only after the flush is successfully finished. This will imply some performance loss, but will ensure durability (the "D" in ACID).
- If your concern is to use an (existing or pristine) MongoDB instance, we encourage you to build the a Kuzzle plugin to flush all the data in it (again, we'll soon provide boilerplate code to make this process easy). Then, if you want to directly access it, you can do it either via your network configuration or by creating a dedicated Kuzzle Controller Plugin. Again, this code would be heavily dependent on your architecture, so we do not provide general plugins for this purpose.

We hope that this answer helps you.
We, of course, welcome your feedback on all the points.

**Comment by farwayer** [comment]
@xbill82 and kuzzle team thanks for the such detailed response! It is the same cool as your documentation. I have some thoughts (and maybe suggestions) about kuzzle database layer.
I will give a full answer later.

**Comment by farwayer** [comment]
I'll start from the abstract things. This should help to understand my point of view. And I'm sorry about my bad english :sweat_smile:  

Kuzzle has real nice features. Events and hooks, build-in auth and permissions, transport agnostic, pub/sub with server filters, client libraries, plugins, scaling and very good code quality can help to make awesome things.

It might give the impression that it is very easy to develop you own backend based on kuzzle. Unfortunately this is not true. Kuzzle is very egocentric. What do I mean? Kuzzle allows you to extend himself but it makes you suffer.

How can you extend some backend frameworks? It's simple in most cases:
```javascript
const server = new FantasticBackendFramework(/*config*/);

server.use(new MyTransportPlugin(/*config*/));
server.use(new MyController1(/*config*/));
server.use(new MyController2(/*config*/));

server.start();
```

What should you do to extend kuzzle with new controller? I didn't trying to create controller yet so steps are from documentation:
  1. `package.json` with plugin options in `pluginInfo` (yes controller must be npm package)
  2. `controllers.js` with configuration (mappings?)
  3. `routes.js` with routes
  4. `controller.js` with controller code
  5. main plugin file (index.js?)
  6. add plugin to kuzzle config

So, what if I need 10 controllers with different behaviour? So much boilerplate code.

Kuzzle looks monolithic from outside. Just like big square with "KUZZLE" from your answer. **You don't use kuzzle but kuzzle uses you code** with many configs you must provide.

Main idea is **Kuzzle should be more flexible for using and extending**. It should not be blackbox running inside docker!
***

Ok, now lets return to databases.

You wrote `ElasticSearch is not meant to be used as a so-called "master" database`. But what do kuzzle keep inside ES? Users and roles. Is it important data? I think so and I **don't want to keep this data inside search engine only**. I can copy it to my own database but **kuzzle will still use ES for auth**. Yes I know I can rewrite passport strategy.
And I am not sure what to do with roles.

What is about writing my own controller to other database? Yes it's possible. And of course I would like to get all kuzzle features. So I need to notify kuzzle about document modifications and send events to plugins. **In result 99% of my controller will be same as your `read` and `write` controllers** except engine calling. And if you will change some internal kitchen (rename `kuzzle.notifier.notifyDocumentCreate` to `kuzzle.notifier.notifyDocumentCreated` for example) I will need to modify my controller.

Can I use client libraries out of the box with my controller? No. `write` and `read` controllers are hardcoded inside code.

Ok, so what I need to use my own database?
  - write `storage controller` (with complex connecting to kuzzle)
  - develop `storage engine`
  - rewrite `passport strategy` (to keep users inside my own database)
  - rewrite `client library(es) `

And as result I will get 3 running databases! Is it all ok?
Maybe ElasticSearch is good database but kuzzle is too hard depends on it.
***

My suggestion is to develop database abstraction layer.

First of all `read` and `write` controllers should be combined  into `storage` controller. They perform one function to provide access to persistent storage. Some actions (like `now` and `serverInfo`) should be moved out from `read` controller to `info` or `server` controller.

In simple case `storage` controller should use some internal component to select *storage backend* based on *collection name*. In more complex cases it should has ability to select backend with dynamic function (just like permission system works).
We can keep part of data in one storage, part in another storage and part in both for example.

Also it need to add new plugin type (`storage` for example) to provide access to different databases. Common interface for such plugins should be designed.

Also it's need to remove `internalStorage` and add ability to choose any database storage.
As option `redis` can also be `storage` plugin and support part of this interface.

Thank you for attention!

**Comment by farwayer** [comment]
I do not claim to truth :smile: But as for be database agnostic please read [this](https://rethinkdb.com/blog/rethinkdb-shutdown/). RethinkDB is not so popular as MongoDB and ES but it is (hope not was) very good product. A lot can happen in this world.

**Comment by xbill82** [comment]
Hi again, @farwayer !
We recently had a meeting and discussed altogether about this thread. As @scottinet said, many hints from your feedback are precious and we might dig them deeply in the near future.
We are currently very busy developing a proper v1.0 as soon as possible, so the investigation concerned in this thread will be postponed to the beginning of v2.0, probably during the first quarter of 2017.

Again, thank you for your feedback and stay tuned :)

**Comment by farwayer** [comment]
I will try to use kuzzle and ES on non-critical project. Imho kuzzle aim for 2.0 is to be more agile and extendable as microcore and core services (db, auth and etc.) as "plugins". Plugins should be injected into core with code or configs. And core should be avaliable as npm package and also. Thank you for response and good luck with releasing!

**Comment by scottinet** [comment]
Believe it or not but this is still an active question. And we're (slowly) progressing toward agnostic storage drivers, allowing kuzzle to work with other storage solutions such as postgres or mongodb. 

As it is now however, there is no point in keeping that issue open: we're still a long way from having such a feature in kuzzle.
